contracts/upkeeps/Cron.sol

```diff
@@ -0,0 +1,255 @@
+import "./UpkeepBase.sol";
+
+/**
+ * @title The Cron contract
+ * @notice A keeper-compatible contract that runs various tasks on cron schedules.
+ * Users must use the encodeCronString() function to encode their cron jobs before
+ * setting them. This keeps all the string manipulation off chain and reduces gas costs.
+ */
+contract Cron is UpkeepBase {
+  //
+  using CronUtility for Spec;
+  using CronUtility for string;
+
+  event CronJobExecuted(uint256 indexed id, uint256 timestamp);
+  event CronJobCreated(uint256 indexed id, address target, bytes handler);
+  event CronJobDeleted(uint256 indexed id);
+
+  error argumentLengthMismatch();
+  error callFailed(uint256 id, string reason);
+  error cronJobIDNotFound(uint256 id);
+
+  struct CronJob {
+    uint256 nextTick;
+    Spec spec;
+    address target;
+    bytes handler;
+  }
+
+  uint256 private s_nextCronJobID = 1;
+  uint256[] private s_activeCronJobIDs;
+  mapping(uint256 => CronJob) private s_cronJobs;
+
+  /**
+   * @param keeperRegistryAddress The address of the keeper registry contract
+   */
+  constructor(address keeperRegistryAddress)
+    UpkeepBase(keeperRegistryAddress)
+  {}
+
+  /**
+   * @notice Get the id of an eligible cron job
+   * @return upkeepNeeded signals if upkeep is needed, performData is an abi encoding
+   * of the id and "next tick" of the elligible cron job
+   */
+  function checkUpkeep(bytes calldata)
+    external
+    view
+    override
+    whenNotPaused()
+    returns (bool, bytes memory)
+  {
+    // DEV: start at a random spot in the list so that checks are
+    // spread evenly among cron jobs
+    uint256 idx;
+    uint256 id;
+    uint256 numCrons = s_activeCronJobIDs.length;
+    uint256 startIdx = block.number % numCrons;
+    CronJob memory cronJob;
+    for (idx = startIdx; idx < numCrons; idx++) {
+      id = s_activeCronJobIDs[idx];
+      cronJob = s_cronJobs[id];
+      if (block.timestamp >= cronJob.nextTick) {
+        return (true, abi.encode(id, cronJob.spec.nextTick()));
+      }
+    }
+    for (idx = 0; idx < startIdx; idx++) {
+      id = s_activeCronJobIDs[idx];
+      cronJob = s_cronJobs[id];
+      if (block.timestamp >= cronJob.nextTick) {
+        return (true, abi.encode(id, cronJob.spec.nextTick()));
+      }
+    }
+    return (false, bytes(""));
+  }
+
+  /**
+   * @notice Executes the cron job with id encoded in performData
+   * @param performData abi encoding of cron job ID and the cron job's next run-at datetime
+   */
+  function performUpkeep(bytes calldata performData)
+    external
+    override
+    onlyKeeperRegistry()
+    whenNotPaused()
+  {
+    (uint256 id, uint256 nextTick) = abi.decode(
+      performData,
+      (uint256, uint256)
+    );
+    CronJob memory cronJob = s_cronJobs[id];
+    if (block.timestamp < cronJob.nextTick) {
+      revert PerformanceNotNeeded();
+    }
+    s_cronJobs[id].nextTick = nextTick;
+    (bool success, bytes memory payload) = cronJob.target.call(cronJob.handler);
```

Comment:
What do you think of making this an abstract contract instead and requiring devs implement a method with their logic instead of calling another contract? This would remove the need to pay for an external call, and worst case they just write that method to make an external call.

Reply:
Alternatively, what if we add `handler` into the `performData`, then we don't have to pay the expensive storage reads when we actually execute `performUpkeep`. To make sure it is encoded correctly, we could record a hash of `handler` and just read and compare that in `performUpkeep`.

Back of the napkin: I think in the minimal case of only a function selector it costs a little more, but if one or more arguments are passed it would save 10k+ gas.

Reply:
I like the 2nd suggestion. Will definitely do that.

Regarding the first one, I viewed the cron contract as a router, where users could add and remove various scheduled tasks as they saw fit (without having to register new upkeeps). If we make this an abstract contract, then I think we loose the ability to support multiple cron jobs with single upkeep contract. Let me know if I'm missing something.

Reply:
Sounds right to me.